render/pixman/renderer.c

```diff
@@ -0,0 +1,365 @@
+	wl_list_remove(&buffer->link);
+	wl_list_remove(&buffer->buffer_destroy.link);
+
+	pixman_image_unref(buffer->image);
+
+	free(buffer);
+}
+
+static struct wlr_pixman_buffer *create_buffer(
+		struct wlr_pixman_renderer *renderer, struct wlr_buffer *wlr_buffer) {
+	struct wlr_pixman_buffer *buffer = calloc(1, sizeof(*buffer));
+	if (buffer == NULL) {
+		wlr_log_errno(WLR_ERROR, "Allocation failed");
+		return NULL;
+	}
+	buffer->buffer = wlr_buffer;
+	buffer->renderer = renderer;
+
+	uint32_t drm_format;
+	struct wlr_dmabuf_attributes dmabuf = {0};
+	struct wlr_shm_attributes shm = {0};
+	if (wlr_buffer_get_dmabuf(wlr_buffer, &dmabuf)) {
+		drm_format = dmabuf.format;
+	} else if (wlr_buffer_get_shm(wlr_buffer, &shm)) {
+		drm_format = shm.format;
+	} else {
+		goto error_buffer;
+	}
+
+	uint32_t format = get_pixman_format_from_drm(drm_format);
+
+	void *data = NULL;
+	size_t stride;
+	if (!buffer_get_data_ptr(wlr_buffer, &data, &stride)) {
+		wlr_log(WLR_ERROR, "Failed to get buffer data");
+		goto error_buffer;
+	}
+
+	wlr_log(WLR_INFO, "width %zu, height %zu", (size_t)wlr_buffer->width,
+			(size_t)wlr_buffer->height);
+	wlr_log(WLR_INFO, "data %p, stride %zu", data, stride);
+
+	buffer->image = pixman_image_create_bits(format, wlr_buffer->width,
+			wlr_buffer->height, data, stride);
+	if (!buffer->image) {
+		wlr_log(WLR_ERROR, "Failed to allocate pixman image\n");
+		goto error_buffer;
+	}
+
+	buffer->buffer_destroy.notify = handle_destroy_buffer;
+	wl_signal_add(&wlr_buffer->events.destroy, &buffer->buffer_destroy);
+
+	wl_list_insert(&renderer->buffers, &buffer->link);
+
+	wlr_log(WLR_DEBUG, "Created pixman buffer %dx%d",
+		wlr_buffer->width, wlr_buffer->height);
+
+	return buffer;
+
+error_buffer:
+	free(buffer);
+	return NULL;
+}
+
+static void pixman_begin(struct wlr_renderer *wlr_renderer, uint32_t width,
+		uint32_t height) {
+	struct wlr_pixman_renderer *renderer = get_renderer(wlr_renderer);
+	renderer->width = width;
+	renderer->height = height;
+}
+
+static void pixman_clear(struct wlr_renderer *wlr_renderer,
+		const float color[static 4]) {
+	struct wlr_pixman_renderer *renderer = get_renderer(wlr_renderer);
+	struct wlr_pixman_buffer *buffer = renderer->current_buffer;
+
+	const struct pixman_color colour = {
+		.red = color[0] * 0xFFFF,
+		.green = color[1] * 0xFFFF,
+		.blue = color[2] * 0xFFFF,
+		.alpha = color[3] * 0xFFFF,
+	};
+
+	pixman_image_t *fill = pixman_image_create_solid_fill(&colour);
+
+	pixman_image_composite32(PIXMAN_OP_SRC, fill, NULL, buffer->image, 0, 0, 0,
+			0, 0, 0, renderer->width, renderer->height);
+
+	pixman_image_unref(fill);
+}
+
+static void pixman_scissor(struct wlr_renderer *wlr_renderer,
+		struct wlr_box *box) {
+	struct wlr_pixman_renderer *renderer = get_renderer(wlr_renderer);
+	struct wlr_pixman_buffer *buffer = renderer->current_buffer;
+
+	if (box != NULL) {
+		struct pixman_region32 region = {0};
+		pixman_region32_init_rect(&region, box->x, box->y, box->width,
```

Comment:
We need to `pixman_region32_fini` after the `pixman_image_set_clip_region32` call, to avoid leaking memory.

Reply:
I think we're good because the `pixman_region32` is allocated on the stack https://cgit.freedesktop.org/pixman/tree/pixman/pixman-region.c#n413

Reply:
`FREE_DATA` hides a `free` call: https://cgit.freedesktop.org/pixman/tree/pixman/pixman-region.c#n216